### ISSUE TITLE
AOT compilation wrong type for host in buffer_t

### ISSUE THREAD

**Issue by guusleijsten** [halide/Halide#1502]
I am currently working on creating a processing pipeline using the AOT compilation approach. I declared an ImageParam as followed:
     ImageParam input(type_of<uint16_t>(), 2);
Obviously I have on the last line of my program:
Added the compile_to_static_library on the bottom of the file and compiled and ran the obtained object file. This gave me halide_pipeline.h.

However now looking at the halide_pipeline.h file I can see that the declaration of buffer_t has the following:
     uint8_t* host;

I get an error because I run the AOT program with an 16-bit image. It says that the types should match. 

Error:
halide_pipeline_run.cpp:70:21: error: cannot convert ‘uint16_t* {aka short unsigned int*}’ to ‘uint8_t* {aka unsigned char*}’ in assignment
     input_buf.host  = &input[0];

Did I miss something? I did also use buffer_t.elem_size = 2, however this doesn't change the fact that I get an type mismatch error.

kind regards, and thanks!

**Comment by guusleijsten** [comment]
I realize that the AOT is too buggy and can not be used for that reason.
The reasons are as follows:
None of the basic examples work. Not the minimal AOT-compiled-example, neither lesson_10 from the tutorial.

My knowledge is unfortunately not sufficient to fix these mistakes myself (it might just be a simple compiler flag that has to be -a instead of -o for example). 

The error message does not make sense.

I will go back to plan B, which is JIT compiled.

**Comment by abadams** [comment]
buffer_t has a single fixed type (it's not templated), which means we had to pick a single type for the host field. Just do a pointer cast.

What error message are you getting?

**Comment by guusleijsten** [comment]
Thankyou for the fast reply. I will comment with some additional information tomorrow.

However I already set up some pretty neat jit_compile alternative. 

It would be nice to get the AOT working so again, I will come back on this tomorrow.

Kind regards

**Comment by guusleijsten** [comment]
As mentioned, I would get back to you on this issue. I first discuss my own case, after that I will show that lesson 10 has the same problem as my case. For the time being, I will use JIT compilation to be able to continue my work. It would be nice however to abstract from the method that I obtain the input file using "buffer_t". 

After I compiled and ran "halide_generate.cpp" I obtain a "gradient.o" and a "gradient.h", I use "compile_to_file" method to obtain these. similarly I compile and run "halide_pipeline.cpp" where I use the "compile_to_static_library" method, the gives me a ".a" and a ".h" file.

Then I run the following command:
$ g++ halide_run.cpp gradient.o -lpthread -o halide_run

I get the following error message, which is for both cases that I described above very similar except for the file names in the message..
Error messages: 
gradient.o: In function `halide_get_symbol':
gradient:(.text.halide_get_symbol+0xe): undefined reference to `dlsym'
gradient.o: In function `halide_load_library':
gradient:(.text.halide_load_library+0xc): undefined reference to `dlopen'
gradient:(.text.halide_load_library+0x19): undefined reference to `dlerror'
gradient.o: In function `halide_get_library_symbol':
gradient:(.text.halide_get_library_symbol+0x6): undefined reference to `dlsym'
collect2: error: ld returned 1 exit status


To be sure that this error keeps popping up, I do the procedure from lesson_10 once more. It gives the following error message, identical to the one above:

lesson_10_halide.a(lesson_10_halide.a.o): In function `halide_get_symbol':
lesson_10_halide:(.text.halide_get_symbol+0xe): undefined reference to `dlsym'
lesson_10_halide.a(lesson_10_halide.a.o): In function `halide_load_library':
lesson_10_halide:(.text.halide_load_library+0xc): undefined reference to `dlopen'
lesson_10_halide:(.text.halide_load_library+0x19): undefined reference to `dlerror'
lesson_10_halide.a(lesson_10_halide.a.o): In function `halide_get_library_symbol':
lesson_10_halide:(.text.halide_get_library_symbol+0x6): undefined reference to `dlsym'
collect2: error: ld returned 1 exit status

I am not sure if it is an issue in the latest release?

Thanks in advance,
Kind regards.
